server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepositoryV2.java

```diff
@@ -0,0 +1,874 @@
+        final Revision firstRevision = primaryRepo.firstRevision();
+        final RevisionRange range = new RevisionRange(firstRevision, firstRevision);
+        return primaryRepo.listCommits(ALL_PATH, 1, range).get(0);
+    }
+
+    @Override
+    public Revision shouldCreateRollingRepository(int minRetentionCommits, int minRetentionDays) {
+        final InternalRepository repo = secondaryRepo != null ? secondaryRepo : primaryRepo;
+        return shouldCreateRollingRepository(repo.headRevision(), minRetentionCommits, minRetentionDays);
+    }
+
+    @Nullable
+    private Revision shouldCreateRollingRepository(Revision headRevision, int minRetentionCommits,
+                                                   int minRetentionDays) {
+        // TODO(minwoox): provide a way to set different minRetentionCommits and minRetentionDays
+        //                in each repository.
+        if ((minRetentionCommits == 0 && minRetentionDays == 0) ||
+            (minRetentionCommits == Integer.MAX_VALUE && minRetentionDays == Integer.MAX_VALUE)) {
+            // Not enabled.
+            return null;
+        }
+
+        final InternalRepository repo = secondaryRepo != null ? secondaryRepo : primaryRepo;
+        if (exceedsMinRetention(repo, headRevision, minRetentionCommits, minRetentionDays)) {
+            return headRevision;
+        }
+        return null;
+    }
+
+    private static boolean exceedsMinRetention(InternalRepository repo, Revision headRevision,
+                                               int minRetentionCommits, int minRetentionDays) {
+        final Revision firstRevision = repo.firstRevision();
+        if (minRetentionCommits != 0) {
+            if (headRevision.major() - firstRevision.major() <= minRetentionCommits) {
+                return false;
+            }
+        }
+
+        if (minRetentionDays != 0) {
+            final Instant secondCommitCreationTime = repo.secondCommitCreationTimeInstant();
+            return secondCommitCreationTime != null &&
+                   secondCommitCreationTime.isBefore(Instant.now().minus(minRetentionDays, ChronoUnit.DAYS));
+        }
+        return true;
+    }
+
+    @Override
+    public void createRollingRepository(Revision rollingRepositoryInitialRevision,
+                                        int minRetentionCommits, int minRetentionDays) {
+        requireNonNull(rollingRepositoryInitialRevision, "rollingRepositoryInitialRevision");
+        checkState(shouldCreateRollingRepository(rollingRepositoryInitialRevision,
+                                                 minRetentionCommits, minRetentionDays) ==
+                   rollingRepositoryInitialRevision, "aaa");
+
+        final InternalRepository secondaryRepo = this.secondaryRepo;
+        if (secondaryRepo != null) {
+            promoteSecondaryRepo(secondaryRepo, rollingRepositoryInitialRevision);
+        } else {
+            createSecondaryRepo(rollingRepositoryInitialRevision);
+        }
```

Comment:
As per SPR, how about removing `createSecondaryRepo()` in `promoteSecondaryRepo()`?
```suggestion
        if (secondaryRepo != null) {
            promoteSecondaryRepo(secondaryRepo, rollingRepositoryInitialRevision);
        }
        createSecondaryRepo(rollingRepositoryInitialRevision);
```